### PR TITLE
Move score to a long plus other tweaks

### PR DESCRIPTION
**Pull request by ConorMcPersonal**
Score now a long, bit-shifted for display
Add a new bread type (Melba)
Customers in shop at the close can still be served (or leave)
GameParameters and BreadBin persist throughout game
Overdue refactor to simplify modifications to GameParameters and focus game.h

### CODE REVIEW

**Comment by mcmenaminadrian** [comment]
I had a chat witha friend who told me gluten free bread would be a good one
to add as it takes forever to toast

File: SplitFiles/base.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+// 1.) Forward declaration: Here is the name of the structure
+// but member-details are omitted.
+struct BreadBinStruct;
+
+// 2.) typedef of the structure
+// Still no details on the members.
+typedef struct BreadBinStruct  BreadBin;
+
+struct BreadTypeStruct {
+    unsigned char     letter;
```
Comment (mcmenaminadrian) [inline SplitFiles/base.h]: and the char, as an 8 bit should probably go last (as a small memory optimisation)

File: SplitFiles/base.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+typedef struct GameComponent_struct  GameComponent;
+
+// 3.) The parameter to the function is listed, using the definition
+// from step 2.)  (note: Still no details on the members yet)
+//typedef void (*GameComponentFunc)(GameComponent* input, GameParameters* params);
+typedef void (*GameComponentFunc)(GameComponent*, GameParameters*);
+
+struct GameParameters_Struct {
+  int       ticks; // How many times has main loop run
+  long       score; // What's the score
```
Comment (mcmenaminadrian) [inline SplitFiles/base.h]: previous comment may have been lost - should put the long as the first entry here as a memory optimisation (avoids fragmentation - possibly)

**Comment by ConorMcPersonal** [comment]
> I had a chat witha friend who told me gluten free bread would be a good one to add as it takes forever to toast

Done

File: SplitFiles/base.h (hunk 2 above)
Comment (ConorMcPersonal) [reply SplitFiles/base.h]: Made both these tweaks, thankfully code now less reliant on parameter order for struct population!

File: SplitFiles/base.h (hunk 1 above)
Comment (ConorMcPersonal) [reply SplitFiles/base.h]: Done

**Review by mcmenaminadrian — COMMENTED** [review]
Not identified the source for this - but: if you burn toast your score still goes up (by a lot in some cases), and the sound never recovers properly and the melba toast faster than everything, not slower!

**Comment by mcmenaminadrian** [comment]
> Not identified the source for this - but: if you burn toast your score still goes up (by a lot in some cases), and the sound never recovers properly and the melba toast faster than everything, not slower!

sorry that's obviously correct for melba, getting confused with gluten, which I am yet to see

**Review by mcmenaminadrian — COMMENTED** [review]
Not convinced the shifting is working here - burn a lot of toast and see what happens - still seems to flip at 32k

**Comment by ConorMcPersonal** [comment]
Bit shifting was never a sinecure, was just going to slow down how fast you could get there.  Think score below zero = goodnight vienna, negative score seem to play havoc with the whole thing (which is what you're seeing too by the sound of it)

**Comment by ConorMcPersonal** [comment]
Gluten free is a low probability order.  melba supposed to be fast.  Think we need to look at scoring and reputation in general - it's currently a bit too generous all round.

File: SplitFiles/game.c [hunk 3]
```diff
@@ -64,14 +63,19 @@ int wait_for_a_key(GameComponent* input, GameParameters* params) {
 }
 
 void tick_func(GameComponent* input, GameParameters* params) { //Now a scoreboard too
-  int last_score = (int)input->ptr;
+  long last_score = (long)input->ptr;
   if (params->score != last_score) {
-    screenNumber(21, 1, params->score);
+    int show_score = (int)(params->score >> 3);
+    screenNumber(21, 1, show_score);
   //  printf(PRINTAT"\x19\x01""%6d", params->score);
```
Comment (mcmenaminadrian) [inline SplitFiles/game.c:70]: I think the issue here is that we have a score like this 0x60000 and shift that to the left by three we get 0xE000 which is still negative. I think I saw a suggestion from you before that a negative score should mean you've lost and that's the way round this and make these things unsigned?

Comment (mcmenaminadrian) [reply SplitFiles/game.c:70]: The other thing, of course, is to stop the score escalating so quickly!

**Pull request merged**